pkg/ccl/importccl/read_import_proc.go

```diff
@@ -469,9 +473,87 @@ func (cp *readImportDataProcessor) doRun(ctx context.Context) error {
 
 	// Sample KVs
 	group.GoCtx(func(ctx context.Context) error {
+		typeBytes := sqlbase.ColumnType{SemanticType: sqlbase.ColumnType_BYTES}
+
 		ctx, span := tracing.ChildSpan(ctx, "sendImportKVs")
 		defer tracing.FinishSpan(span)
 
+		if cp.spec.IngestDirectly {
+			// TODO(dt): allow configuring these.
+			const sortBatchSize = 1000000
+			// TODO(dt): track byte size as well as key count to trigger flushes.
+
+			// TODO(dt): buffer to disk instead of all in-mem.
+
+			writeTS := hlc.Timestamp{WallTime: cp.spec.WalltimeNanos}
+			adder, err := cp.flowCtx.BulkAdder(ctx, cp.flowCtx.ClientDB, 32<<20, writeTS)
+			if err != nil {
+				return err
+			}
+			defer adder.Close()
+
+			buf := make(roachpb.KeyValueByKey, 0, sortBatchSize)
+
+			for kvBatch := range kvCh {
+
+				// TODO(dt): consider checking if we want to flush an almost-full buffer
+				// to make room for a large batch.
+				buf = append(buf, kvBatch...)
+
+				if len(buf) > sortBatchSize {
+					sort.Sort(buf)
```

Comment:
This inner block could be moved to a common func so it shares impl code with the same block below.